src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs

```diff
@@ -0,0 +1,283 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.Cli;
+using Microsoft.DotNet.Cli.NuGetPackageInstaller;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ToolPackage;
+using Microsoft.NET.Sdk.WorkloadManifestReader;
+using NuGet.Versioning;
+using static Microsoft.NET.Sdk.WorkloadManifestReader.WorkloadResolver;
+using EnvironmentProvider = Microsoft.DotNet.NativeWrapper.EnvironmentProvider;
+
+namespace Microsoft.DotNet.Workloads.Workload.Install
+{
+    internal class NetSdkManagedInstaller : PackWorkloadInstallerBase
+    {
+        private readonly IReporter _reporter;
+        private readonly string _workloadMetadataDir;
+        private readonly string _installedWorkloadDir = "InstalledWorkloads";
+        private readonly string _installedPacksDir = "InstalledPacks";
+        protected readonly string _dotnetDir;
+        protected readonly string _tempPackagesDir;
+        private INuGetPackageInstaller _nugetPackageInstaller;
+
+        public NetSdkManagedInstaller(
+            IReporter reporter,
+            INuGetPackageInstaller nugetPackageInstaller = null,
+            string dotnetDir =  null)
+        {
+            _dotnetDir = dotnetDir ?? EnvironmentProvider.GetDotnetExeDirectory();
+            _tempPackagesDir = Path.Combine(_dotnetDir, "metadata", "temp");
+            _nugetPackageInstaller = nugetPackageInstaller ?? new NuGetPackageInstaller(_tempPackagesDir, sourceUrl: "https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json");
+            _workloadMetadataDir = Path.Combine(_dotnetDir, "metadata", "workloads");
+            _reporter = reporter;
+        }
+
+        public override void InstallWorkloadPack(PackInfo packInfo, string featureBand, bool useOfflineCache = false)
+        {
+            if (useOfflineCache)
+            {
+                throw new NotImplementedException();
+            }
+
+            _reporter.WriteLine(string.Format(LocalizableStrings.InstallingPackVersionMessage, packInfo.Id, packInfo.Version));
+            var tempsToDelete = new List<string>();
+            try
+            {
+                TransactionalAction.Run(
+                    action: () =>
+                    {
+                        if (!PackIsInstalled(packInfo))
+                        {
+                            var packPath = _nugetPackageInstaller.InstallPackageAsync(new PackageId(packInfo.Id), new NuGetVersion(packInfo.Version)).Result;
+                            tempsToDelete.Add(packPath);
+                            if (packInfo.Kind.Equals(WorkloadPackKind.Library) || packInfo.Kind.Equals(WorkloadPackKind.Template))
+                            {
+                                Directory.CreateDirectory(Path.GetDirectoryName(packInfo.Path));
+                                File.Copy(packPath, packInfo.Path);
```

Comment:
As we discussed in the meeting, a more robust way is to put everything in a temp directory first and then run a FileAccessRetrier.RetryOnMoveAccessFailure(() => Directory.Move

If the operation fails in the halfway. The folder will be less of a headache (be there as a whole or not at all)